server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java

```diff
@@ -294,7 +348,105 @@ protected MapperMergeContext createChildContext(MapperMergeContext mapperMergeCo
 
     @Override
     public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
-        // IgnoredSourceFieldMapper integration takes care of writing the source for nested objects.
-        return SourceLoader.SyntheticFieldLoader.NOTHING;
+        SourceLoader sourceLoader = new SourceLoader.Synthetic(() -> super.syntheticFieldLoader(mappers.values().stream(), true), NOOP);
+        var storedFieldsLoader = StoredFieldLoader.create(false, sourceLoader.requiredStoredFields());
+        return new NestedFieldLoader(
+            storedFieldsLoader,
+            sourceLoader,
+            () -> bitSetProducer.apply(parentNestedTypeFilter),
+            nestedTypeFilter
+        );
+    }
+
+    private class NestedFieldLoader implements SourceLoader.SyntheticFieldLoader {
+        private final org.elasticsearch.index.fieldvisitor.StoredFieldLoader storedFieldsLoader;
+        private final SourceLoader sourceLoader;
+        private final Supplier<BitSetProducer> parentBitSetProducer;
+        private final Query childFilter;
+
+        private List<Integer> children;
+        private LeafStoredFieldLoader leafStoredFields;
+        private SourceLoader.Leaf leafSource;
+
+        private NestedFieldLoader(
+            org.elasticsearch.index.fieldvisitor.StoredFieldLoader storedFieldsLoader,
+            SourceLoader sourceLoader,
+            Supplier<BitSetProducer> parentBitSetProducer,
+            Query childFilter
+        ) {
+            this.storedFieldsLoader = storedFieldsLoader;
+            this.sourceLoader = sourceLoader;
+            this.parentBitSetProducer = parentBitSetProducer;
+            this.childFilter = childFilter;
+        }
+
+        @Override
+        public Stream<Map.Entry<String, StoredFieldLoader>> storedFieldLoaders() {
+            return Stream.of();
+        }
+
+        @Override
+        public DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf) throws IOException {
+            this.children = null;
+            this.leafStoredFields = storedFieldsLoader.getLoader(leafReader.getContext(), null);
```

Comment:
This is bypassing the way how stored fields are loaded for synthesizing the _source. (The `storedFieldLoaders()` method returns an empty stream. ) I don't see a test for synthesizing _source with nested field type and leaf field that is stored only, or did I miss it? I don't fully follow yet how this would work.

Reply:
The stored field loader is configured [here](https://github.com/elastic/elasticsearch/blob/315d3c4a5bf72ad537c4b7a0dbb5c57ef42ae33e/server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java#L352)
It is responsible for loading the nested stored fields on the child documents and that includes the ignored source field since they are now available at this level. 
There's a test for a disabled object field under a nested field [here](https://github.com/elastic/elasticsearch/blob/64c7dfe2d2b250daab4492824a83cae8af51e56e/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.create/20_synthetic_source.yml#L931). 
I also added an explicit [test](https://github.com/elastic/elasticsearch/blob/64c7dfe2d2b250daab4492824a83cae8af51e56e/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.create/20_synthetic_source.yml#L1046) for a stored text field.